posts/saving-and-analyzing-trending-topics-on-twitter-using-aws-athena-lambda-and-cdk.md

```diff
@@ -0,0 +1,198 @@
+    body = '\n'.join(list(map(lambda trend: json.dumps(trend._json), trends)))
+
+    # Save to S3 with a key like "2020-07-23 05:25:08"
+    key = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')
+    s3.Bucket(os.environ['BUCKET']).put_object(Key=key, Body=body)
+
+    return {
+        'message': 'UPLOAD OK'
+    }
+```
+
+## Stack
+
+The benefits of CDK is that you can define your whole infrastructure with few lines of code and have all the permissions configured. Below you have the code that create both the S3 bucket, the lambda and set a rule to execute the lambda every 15mins. The lambda's environment variables are used to pass the credentials, the S3 bucket and the WOEID (the location used to get the trending topics from). New York is used for this blog but you an replace the WOEID with one of your choice, [here is the list](https://nations24.com/world-wide) of what is available.
+
+```python
+from aws_cdk import (
+    aws_s3 as s3,
+    aws_events as events,
+    aws_lambda as lambda_,
+    aws_events_targets as targets,
+    core,
+)
+
+
+class TwitterTrendsStack(core.Stack):
+
+    def __init__(self, scope: core.Construct, id: str, **kwargs) -> None:
+        super().__init__(scope, id, **kwargs)
+
+        bucket = s3.Bucket(self, 'tlebrun-playground')
+
+        handler = lambda_.Function(self, 'TrendsHandler',
+                                   runtime=lambda_.Runtime.PYTHON_3_8,
+                                   code=lambda_.Code.asset('resources'),
+                                   handler='trends.main',
+                                   environment=dict(BUCKET=bucket.bucket_name,
+                                                    CONSUMER_KEY='',
+                                                    CONSUMER_SECRET='',
+                                                    ACCESS_TOKEN_KEY='',
+                                                    ACCESS_TOKEN_SECRET='',
+                                                    WOEID='2459115')
+                                   )
+
+        bucket.grant_read_write(handler)
+
+        events.Rule(self, 'RuleHandler',
+                    rule_name='Twitter_Trends_15_Minutes_Rule',
+                    schedule=events.Schedule.rate(duration=core.Duration.minutes(15)),
+                    targets=[targets.LambdaFunction(handler)])
+```
+
+The project is available at this [GitHub repository](https://github.com/Falydoor/twitter-trends) in case you want to clone it.
+
+# Data Visualization with AWS Athena
+
+## Database and table creation
+
+Once you have the lambda running for few days, you will be able to view the data in few minutes using [AWS Athena](https://aws.amazon.com/athena/). Athena lets you query your data stored on S3 without the burden of setting up a whole database and having batch processes running. You can create a new database using the AWS Glue service using the console and then create the table using the Athena service.
```

Comment:
Athena lets you query your data stored on S3 without having to set up an entire database and having batch processes running.